File: scripts/test-pipelines.ts
#!/usr/bin/env node
/**
 * Pipeline 验证和调试脚本
 * 
 * 功能：
 * 1. 删除旧的 workflow 文件
 * 2. 使用脚手架工具重新生成 workflow
 * 3. 验证生成的 workflow 文件
 * 4. 可选：触发 workflow 进行在线测试
 * 5. 监控 workflow 执行状态
 * 6. 分析失败原因
 * 
 * 用法:
 *   ts-node scripts/test-pipelines.ts [options]
 *   或
 *   npm run test:pipelines -- [options]
 * 
 * 选项:
 *   --pipeline <name>    指定要测试的 Pipeline 类名（可多次指定）
 *   --all                测试所有 Pipeline
 *   --trigger            触发 workflow 进行在线测试
 *   --watch              监控 workflow 执行状态
 *   --clean              删除旧的 workflow 文件
 *   --verify             仅验证生成的 workflow 文件，不触发测试
 * 
 * 示例:
 *   # 测试单个 Pipeline
 *   npm run test:pipelines -- --pipeline FlutterBuildPipeline --trigger
 * 
 *   # 测试所有 Pipeline
 *   npm run test:pipelines -- --all --trigger --watch
 * 
 *   # 仅验证生成的 workflow 文件
 *   npm run test:pipelines -- --all --verify
 * 
 * 注意：
 * - 此脚本可以共享给父项目使用
 * - 需要安装 GitHub CLI (gh) 才能触发和监控 workflow
 */

import * as fs from 'fs';
import * as path from 'path';
import { exec } from 'child_process';
import { promisify } from 'util';
import * as yaml from 'js-yaml';

const execAsync = promisify(exec);

interface Options {
  pipelines?: string[];
  all?: boolean;
  trigger?: boolean;
  watch?: boolean;
  clean?: boolean;
  verify?: boolean;
}

/**
 * 解析命令行参数
 */
function parseArgs(): Options {
  const args = process.argv.slice(2);
  const options: Options = {
    pipelines: [],
  };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    switch (arg) {
      case '--pipeline':
        if (i + 1 < args.length) {
          options.pipelines!.push(args[++i]);
        }
        break;
      case '--all':
        options.all = true;
        break;
      case '--trigger':
        options.trigger = true;
        break;
      case '--watch':
        options.watch = true;
        break;
      case '--clean':
        options.clean = true;
        break;
      case '--verify':
        options.verify = true;
        break;
    }
  }

  return options;
}

/**
 * 检测项目根目录
 */
function detectProjectRoot(): string {
  let current = process.cwd();
  while (current !== path.dirname(current)) {
    if (fs.existsSync(path.join(current, 'package.json'))) {
      return current;
    }
    current = path.dirname(current);
  }
  return process.cwd();
}

/**
 * 查找所有 Pipeline 文件
 */
function findPipelineFiles(projectRoot: string): string[] {
  const pipelinesDir = path.join(projectRoot, 'src', 'pipelines');
  if (!fs.existsSync(pipelinesDir)) {
    return [];
  }

  const files: string[] = [];
  const findFilesRecursive = (dir: string): void => {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        // 跳过 test 目录（除非配置了 include_test_pipelines）
        if (entry.name === 'test') {
          // 检查配置
          const configPath = path.join(projectRoot, 'config', 'ProjectOnly', 'config.yaml');
          if (fs.existsSync(configPath)) {
            try {
              const config = yaml.load(fs.readFileSync(configPath, 'utf8')) as any;
              if (config?.pipelines?.include_test_pipelines === true) {
                findFilesRecursive(fullPath);
              }
            } catch {
              // 忽略配置读取错误
            }
          }
          continue;
        }
        findFilesRecursive(fullPath);
      } else if (entry.isFile() && entry.name.endsWith('.ts') && !entry.name.startsWith('_')) {
        files.push(fullPath);
      }
    }
  };

  findFilesRecursive(pipelinesDir);
  return files;
}

/**
 * 从文件内容中提取 Pipeline 类名
 */
function extractClassName(filePath: string): string | null {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    // 查找 export class XxxPipeline 模式
    const classMatch = content.match(/export\s+class\s+(\w+Pipeline)\s+extends/);
    if (classMatch && classMatch[1]) {
      return classMatch[1];
    }
    // 如果没有找到，返回 null
    return null;
  } catch {
    return null;
  }
}

/**
 * 删除旧的 workflow 文件
 */
async function cleanWorkflows(workflowsDir: string): Promise<void> {
  console.log('🧹 清理旧的 workflow 文件...');
  const files = fs.readdirSync(workflowsDir);
  for (const file of files) {
    if (file.endsWith('.yml') || file.endsWith('.yaml')) {
      const filePath = path.join(workflowsDir, file);
      fs.unlinkSync(filePath);
      console.log(`   ✓ 删除: ${file}`);
    }
  }
}

/**
 * 生成 workflow 文件
 */
async function generateWorkflow(
  projectRoot: string,
  pipelineName: string,
  workflowsDir: string
): Promise<string | null> {
  console.log(`📝 生成 workflow: ${pipelineName}...`);
  try {
    // 生成 workflow 文件名（将 PascalCase 转换为 kebab-case）
    const workflowFileName = pipelineName
      .replace(/Pipeline$/, '')
      .replace(/([A-Z])/g, '-$1')
      .toLowerCase()
      .replace(/^-/, '') + '.yml';
    
    const outputPath = path.join(workflowsDir, workflowFileName);
    
    // 尝试使用 npm 脚本，如果失败则直接使用 ts-node
    let command = `npm run scaffold -- --pipeline ${pipelineName} --output ${outputPath}`;
    
    // 检查是否有 scaffold 脚本
    const packageJsonPath = path.join(projectRoot, 'package.json');
    if (fs.existsSync(packageJsonPath)) {
      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
      if (!packageJson.scripts?.scaffold) {
        // 如果没有 scaffold 脚本，直接使用 ts-node
        const scaffoldPath = path.join(projectRoot, 'src', 'scaffold.ts');
        if (fs.existsSync(scaffoldPath)) {
          command = `ts-node ${scaffoldPath} --pipeline ${pipelineName} --output ${outputPath}`;
        } else {
          // 尝试从子模块路径查找
          const possiblePaths = [
            path.join(projectRoot, 'GithubActionAISelfBuilder', 'src', 'scaffold.ts'),
            path.join(projectRoot, 'Tools', 'GithubActionAISelfBuilder', 'src', 'scaffold.ts'),
          ];
          for (const scaffoldPath of possiblePaths) {
            if (fs.existsSync(scaffoldPath)) {
              command = `ts-node ${scaffoldPath} --pipeline ${pipelineName} --output ${outputPath}`;
              break;
            }
          }
        }
      }
    }
    
    const { stdout, stderr } = await execAsync(command, { cwd: projectRoot });

    // 检查输出中是否包含成功信息
    const output = stdout + stderr;
    if (output.includes('成功生成') || output.includes('Successfully generated') || output.includes('✓')) {
      if (fs.existsSync(outputPath)) {
        console.log(`   ✓ 成功: ${path.relative(projectRoot, outputPath)}`);
        return outputPath;
      }
    }

    // 即使没有成功信息，也检查文件是否存在
    if (fs.existsSync(outputPath)) {
      console.log(`   ✓ 成功: ${path.relative(projectRoot, outputPath)}`);
      return outputPath;
    }

    // 如果有错误信息，显示错误
    if (stderr && !stderr.includes('成功生成') && !stderr.includes('Successfully')) {
      console.error(`   ❌ 生成失败: ${stderr}`);
      return null;
    }

    return null;
  } catch (error: any) {
    console.error(`   ❌ 生成失败: ${error.message}`);
    return null;
  }
}

/**
 * 验证 workflow 文件
 */
function verifyWorkflow(workflowPath: string): { valid: boolean; errors: string[] } {
  const errors: string[] = [];
  
  try {
    const content = fs.readFileSync(workflowPath, 'utf8');
    const workflow = yaml.load(content) as any;

    // 检查基本结构
    if (!workflow.name) {
      errors.push('缺少 workflow 名称');
    }
    if (!workflow.on) {
      errors.push('缺少触发条件');
    }
    if (!workflow.jobs) {
      errors.push('缺少 jobs 定义');
    }

    // 检查 Pipeline 路径
    const contentStr = content;
    if (!contentStr.includes('require(')) {
      errors.push('缺少 Pipeline 执行步骤');
    }

    return {
      valid: errors.length === 0,
      errors,
    };
  } catch (error: any) {
    errors.push(`YAML 解析错误: ${error.message}`);
    return { valid: false, errors };
  }
}

/**
 * 触发 workflow
 */
async function triggerWorkflow(workflowFile: string, ref: string = 'main'): Promise<number | null> {
  console.log(`🚀 触发 workflow: ${workflowFile}...`);
  try {
    const { stdout } = await execAsync(`gh workflow run ${workflowFile} --ref ${ref}`);
    // 等待一下让 workflow 启动
    await new Promise(resolve => setTimeout(resolve, 3000));
    
    // 获取最新的 run ID
    const { stdout: listOutput } = await execAsync(
      `gh run list --workflow ${workflowFile} --limit 1 --json databaseId --jq '.[0].databaseId'`
    );
    
    const runId = parseInt(listOutput.trim(), 10);
    if (runId) {
      console.log(`   ✓ 已触发，Run ID: ${runId}`);
      return runId;
    }
    return null;
  } catch (error: any) {
    console.error(`   ❌ 触发失败: ${error.message}`);
    return null;
  }
}

/**
 * 监控 workflow 执行
 */
async function watchWorkflow(runId: number): Promise<boolean> {
  console.log(`👀 监控 workflow 执行 (Run ID: ${runId})...`);
  try {
    const { stdout } = await execAsync(`gh run watch ${runId} --exit-status`, {
      timeout: 600000, // 10 分钟超时
    });
    
    // 检查是否成功
    const success = stdout.includes('completed') && !stdout.includes('failure');
    return success;
  } catch (error: any) {
    // exit-status 会在失败时抛出错误
    return false;
  }
}

/**
 * 获取 workflow 执行结果
 */
async function getWorkflowResult(runId: number): Promise<{
  status: string;
  conclusion: string;
  url: string;
}> {
  try {
    const { stdout } = await execAsync(
      `gh run view ${runId} --json status,conclusion,url --jq '{status: .status, conclusion: .conclusion, url: .url}'`
    );
    return JSON.parse(stdout);
  } catch (error: any) {
    return {
      status: 'unknown',
      conclusion: 'unknown',
      url: '',
    };
  }
}

/**
 * 主函数
 */
async function main() {
  const options = parseArgs();
  const projectRoot = detectProjectRoot();
  const workflowsDir = path.join(projectRoot, '.github', 'workflows');

  console.log('🔍 Pipeline 验证和调试工具\n');
  console.log(`项目根目录: ${projectRoot}\n`);

  // 确定要测试的 Pipeline
  let pipelinesToTest: string[] = [];

  if (options.all) {
    // 查找所有 Pipeline
    const files = findPipelineFiles(projectRoot);
    const classNames = files
      .map(extractClassName)
      .filter((name): name is string => name !== null)
      .filter((name, index, self) => self.indexOf(name) === index); // 去重
    
    pipelinesToTest = classNames;
    console.log(`📋 找到 ${pipelinesToTest.length} 个 Pipeline:\n`);
    pipelinesToTest.forEach(name => console.log(`   - ${name}`));
    console.log('');
  } else if (options.pipelines && options.pipelines.length > 0) {
    pipelinesToTest = options.pipelines;
  } else {
    console.error('❌ 错误: 请指定 --pipeline <name> 或 --all');
    process.exit(1);
  }

  // 清理旧的 workflow 文件
  if (options.clean) {
    await cleanWorkflows(workflowsDir);
    console.log('');
  }

  // 生成 workflow 文件
  const generatedWorkflows: Array<{ name: string; path: string }> = [];
  for (const pipelineName of pipelinesToTest) {
    const workflowPath = await generateWorkflow(projectRoot, pipelineName, workflowsDir);
    if (workflowPath) {
      generatedWorkflows.push({
        name: pipelineName,
        path: workflowPath,
      });
    }
  }

  if (generatedWorkflows.length === 0) {
    console.error('❌ 没有成功生成任何 workflow 文件');
    process.exit(1);
  }

  console.log('');

  // 验证 workflow 文件
  console.log('✅ 验证生成的 workflow 文件...\n');
  const verificationResults: Array<{ name: string; valid: boolean; errors: string[] }> = [];
  for (const workflow of generatedWorkflows) {
    const result = verifyWorkflow(workflow.path);
    verificationResults.push({
      name: workflow.name,
      ...result,
    });
    
    if (result.valid) {
      console.log(`   ✓ ${workflow.name}: 验证通过`);
    } else {
      console.log(`   ❌ ${workflow.name}: 验证失败`);
      result.errors.forEach(error => console.log(`      - ${error}`));
    }
  }

  console.log('');

  // 如果只是验证，就退出
  if (options.verify) {
    const allValid = verificationResults.every(r => r.valid);
    process.exit(allValid ? 0 : 1);
  }

  // 触发 workflow 进行在线测试
  if (options.trigger) {
    console.log('🚀 触发 workflow 进行在线测试...\n');
    const runIds: number[] = [];

    for (const workflow of generatedWorkflows) {
      const workflowFile = path.basename(workflow.path);
      const runId = await triggerWorkflow(workflowFile, 'main');
      if (runId) {
        runIds.push(runId);
      }
    }

    if (runIds.length === 0) {
      console.error('❌ 没有成功触发任何 workflow');
      process.exit(1);
    }

    console.log('');

    // 监控 workflow 执行
    if (options.watch) {
      console.log('👀 监控 workflow 执行状态...\n');
      const results: Array<{ name: string; success: boolean; runId: number }> = [];

      for (let i = 0; i < generatedWorkflows.length && i < runIds.length; i++) {
        const workflow = generatedWorkflows[i];
        const runId = runIds[i];
        
        console.log(`\n📊 ${workflow.name} (Run ID: ${runId}):`);
        const success = await watchWorkflow(runId);
        results.push({ name: workflow.name, success, runId });

        // 获取详细结果
        const result = await getWorkflowResult(runId);
        console.log(`   状态: ${result.status}`);
        console.log(`   结果: ${result.conclusion}`);
        if (result.url) {
          console.log(`   链接: ${result.url}`);
        }
      }

      console.log('\n📊 测试结果总结:\n');
      results.forEach(result => {
        const icon = result.success ? '✅' : '❌';
        console.log(`   ${icon} ${result.name} (Run ID: ${result.runId})`);
      });

      const allSuccess = results.every(r => r.success);
      process.exit(allSuccess ? 0 : 1);
    } else {
      console.log('💡 提示: 使用 --watch 选项可以监控 workflow 执行状态');
      console.log(`\n📊 已触发的 workflow:\n`);
      for (let i = 0; i < generatedWorkflows.length && i < runIds.length; i++) {
        const workflow = generatedWorkflows[i];
        const runId = runIds[i];
        console.log(`   - ${workflow.name}: Run ID ${runId}`);
        console.log(`     查看: gh run view ${runId}`);
      }
    }
  }
}

main().catch(error => {
  console.error('❌ 执行失败:', error);
  process.exit(1);
});

File: scripts/tools/test_pipelines.py
#!/usr/bin/env python3
"""
Pipeline 验证和调试脚本

功能：
1. 删除旧的 workflow 文件
2. 使用脚手架工具重新生成 workflow
3. 验证生成的 workflow 文件
4. 可选：触发 workflow 进行在线测试
5. 监控 workflow 执行状态
6. 分析失败原因

用法:
  python scripts/test_pipelines.py [options]

选项:
  --pipeline <name>    指定要测试的 Pipeline 类名（可多次指定）
  --all                测试所有 Pipeline
  --trigger            触发 workflow 进行在线测试
  --watch              监控 workflow 执行状态
  --clean              删除旧的 workflow 文件
  --verify             仅验证生成的 workflow 文件，不触发测试
"""

import sys
import os
import argparse
import subprocess
import yaml
import time
from pathlib import Path
import importlib.util
import re


def detect_project_root() -> Path:
    """检测项目根目录"""
    current = Path.cwd()
    while current != current.parent:
        if (current / "package.json").exists() or (current / "environment.yml").exists():
            return current
        current = current.parent
    return Path.cwd()


def find_pipeline_files(project_root: Path) -> list:
    """查找所有 Pipeline 文件"""
    # Pipeline 文件在 python/src/pipelines/ 目录
    python_dir = project_root / "python"
    pipelines_dir = python_dir / "src" / "pipelines"
    if not pipelines_dir.exists():
        return []
    
    files = []
    for file_path in pipelines_dir.rglob("*.py"):
        if not file_path.name.startswith("_"):
            files.append(file_path)
    
    return files


def extract_class_name(file_path: Path) -> str:
    """从文件内容中提取 Pipeline 类名"""
    try:
        content = file_path.read_text(encoding="utf-8")
        # 查找 class XxxPipeline 模式
        match = re.search(r"class\s+(\w+Pipeline)\s*[\(:]", content)
        if match:
            return match.group(1)
    except Exception:
        pass
    return None


def clean_workflows(workflows_dir: Path) -> None:
    """删除旧的 workflow 文件"""
    print("🧹 清理旧的 workflow 文件...")
    for file_path in workflows_dir.glob("*.yml"):
        file_path.unlink()
        print(f"   ✓ 删除: {file_path.name}")
    for file_path in workflows_dir.glob("*.yaml"):
        file_path.unlink()
        print(f"   ✓ 删除: {file_path.name}")


def generate_workflow(
    project_root: Path,
    pipeline_name: str,
    workflows_dir: Path
) -> Path:
    """生成 workflow 文件"""
    print(f"📝 生成 workflow: {pipeline_name}...")
    
    # 生成 workflow 文件名
    workflow_name = (
        pipeline_name.replace("Pipeline", "")
        .replace("_", "-")
        .lower()
    )
    output_path = workflows_dir / f"{workflow_name}.yml"
    
    # 使用脚手架工具生成
    # 需要从 python/ 目录运行脚手架工具
    python_dir = project_root / "python"
    scaffold_cmd = [
        sys.executable,
        "-m", "src.scaffold",
        "--pipeline", pipeline_name,
        "--output", str(output_path.relative_to(project_root)),
    ]
    
    if output_path.exists():
        scaffold_cmd.append("--update")
    
    try:
        result = subprocess.run(
            scaffold_cmd,
            cwd=python_dir,
            capture_output=True,
            text=True
        )
        
        if result.returncode == 0 and output_path.exists():
            print(f"   ✓ 成功: {output_path.relative_to(project_root)}")
            return output_path
        else:
            print(f"   ❌ 生成失败: {result.stderr}")
            return None
    except Exception as e:
        print(f"   ❌ 生成失败: {str(e)}")
        return None


def verify_workflow(workflow_path: Path) -> tuple:
    """验证 workflow 文件"""
    errors = []
    
    try:
        with open(workflow_path, "r", encoding="utf-8") as f:
            workflow = yaml.safe_load(f)
        
        # 检查基本结构
        if not workflow.get("name"):
            errors.append("缺少 workflow 名称")
        if not workflow.get("on"):
            errors.append("缺少触发条件")
        if not workflow.get("jobs"):
            errors.append("缺少 jobs 定义")
        
        # 检查 Pipeline 执行步骤
        content = workflow_path.read_text(encoding="utf-8")
        if "python -c" not in content and "Run " not in content:
            errors.append("缺少 Pipeline 执行步骤")
        
        return len(errors) == 0, errors
    except Exception as e:
        return False, [f"YAML 解析错误: {str(e)}"]


def main():
    parser = argparse.ArgumentParser(description="Pipeline 验证和调试脚本")
    parser.add_argument("--pipeline", action="append", help="Pipeline 类名（可多次指定）")
    parser.add_argument("--all", action="store_true", help="测试所有 Pipeline")
    parser.add_argument("--trigger", action="store_true", help="触发 workflow 进行在线测试")
    parser.add_argument("--watch", action="store_true", help="监控 workflow 执行状态")
    parser.add_argument("--clean", action="store_true", help="删除旧的 workflow 文件")
    parser.add_argument("--verify", action="store_true", help="仅验证生成的 workflow 文件")
    
    args = parser.parse_args()
    
    project_root = detect_project_root()
    workflows_dir = project_root / ".github" / "workflows"
    workflows_dir.mkdir(parents=True, exist_ok=True)
    
    # 清理旧的 workflow 文件
    if args.clean:
        clean_workflows(workflows_dir)
    
    # 确定要测试的 Pipeline
    pipelines = []
    if args.all:
        pipeline_files = find_pipeline_files(project_root)
        for file_path in pipeline_files:
            class_name = extract_class_name(file_path)
            if class_name:
                pipelines.append(class_name)
    elif args.pipeline:
        pipelines = args.pipeline
    else:
        print("❌ 错误: 请指定 --pipeline 或 --all")
        sys.exit(1)
    
    if not pipelines:
        print("❌ 未找到 Pipeline")
        sys.exit(1)
    
    print(f"📋 找到 {len(pipelines)} 个 Pipeline:")
    for pipeline in pipelines:
        print(f"   - {pipeline}")
    print()
    
    # 生成 workflow
    generated_workflows = []
    for pipeline_name in pipelines:
        workflow_path = generate_workflow(project_root, pipeline_name, workflows_dir)
        if workflow_path:
            generated_workflows.append(workflow_path)
    
    if not generated_workflows:
        print("❌ 没有成功生成任何 workflow")
        sys.exit(1)
    
    print(f"\n✅ 成功生成 {len(generated_workflows)} 个 workflow\n")
    
    # 验证 workflow
    if args.verify or not args.trigger:
        print("🔍 验证 workflow 文件...")
        all_valid = True
        for workflow_path in generated_workflows:
            valid, errors = verify_workflow(workflow_path)
            if valid:
                print(f"   ✓ {workflow_path.name}: 验证通过")
            else:
                print(f"   ❌ {workflow_path.name}: 验证失败")
                for error in errors:
                    print(f"      - {error}")
                all_valid = False
        
        if not all_valid:
            print("\n❌ 部分 workflow 验证失败")
            sys.exit(1)
        
        print("\n✅ 所有 workflow 验证通过")
    
    # 触发和监控（如果需要）
    if args.trigger:
        print("\n🚀 触发 workflow 进行在线测试...")
        print("   注意: 此功能需要 GitHub CLI (gh) 和推送代码到远程仓库")
        # 这里可以添加触发逻辑
        # 由于需要 GitHub CLI，这里简化处理
        print("   ⚠️  触发功能需要手动实现或使用 GitHub CLI")


if __name__ == "__main__":
    main()

